### DISCUSSION TITLE
Add `DATEPART` to expression language

### DISCUSSION THREAD

**Discussion by nblumhardt** [datalust/seq-tickets#1108] [Ideas]
Would allow for queries like "only events on weekdays".

Via: https://twitter.com/firstdrafthell/status/940465106021965824

**Comment by chessels** [comment]
What about other date parts, like _year_, _month_, _week_, etc?

**Comment by nblumhardt** [comment]
Thanks for the suggestion @chessels, yes, these seem reasonable to consider under this ticket :+1:

**Comment by skovsende** [comment]
Any progress on this?

**Comment by nblumhardt** [comment]
Still on the radar, but no progress yet - thanks for the nudge @skovsende :+1:

**Comment by CrossBound** [comment]
I am looking for a way to say something simple like '@Timestamp >= Today()' which would give me any events occurring at or after 00:00 UTC today. Is there a way to do this currently or is this ticket the solution to that as well? I know there is a Now() function and I could do Now() - 24h but I'm looking for more of a "today" functionality than a rolling 24 hours.

**Comment by nblumhardt** [comment]
Hi @CrossBound - I think this should work, though it's a little hacky:

```sql
@Timestamp > Now() - TimeOfDay(Now(), -7)
```

where `-7` is your time zone offset, in hours. Let me know if this helps!

**Comment by CrossBound** [comment]
Is there currently any functions that return the current timezone offset? If not that makes these type of queries difficult because you have to change them every time we go through daylight savings time change.

**Comment by nblumhardt** [comment]
Unfortunately, no - I think the major stumbling block is that the offset depends on the client, not the server. We might be able to do something like `OffsetOf('PST')` 🤔

**Comment by nblumhardt** [comment]
Hi folks! 👋 

This one's been a long time in the making - but we're mere days away from 2021.2, and `datepart()`, along with `offsetin()` are going to make the cut 😊 🎉🎉🎉

Did the event occur on a Wednesday?

```js
datepart(@Timestamp, 'weekday', offsetin('Australia/Brisbane', @Timestamp)) = 3
```

In June?

```js
datepart(@Timestamp, 'month', offsetin('Australia/Brisbane', @Timestamp)) = 6
```

`offsetin()` accepts an [IANA time zone name](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), and because of DST, also needs a point in time at which to compute the offset. It's currently a bit verbose, but covers just about everything we can think of, and we hope to add a little syntax sugar in the future.

Supported "parts" are: `date`, `time`, `year`, `month`, `day`, `hour`, `minute`, `second`, and `weekday`.

The `date` and `time` values are ticks-since-0001-01-01, and ticks, respectively. Calendar/clock time components are the usual integers - 2021 for `year`, etc., and `weekday` is a zero-based day number starting from Sunday = 0.

Here's how @CrossBound's "timestamp is today" query looks:

```js
datepart(@Timestamp, 'date', offsetin('Australia/Brisbane', @Timestamp)) =
  datepart(now(), 'date', offsetin('Australia/Brisbane', now())
```

Two important things to note:

 * the `part` argument to `datepart` is a string, e.g. `'day'`, not an identifier as you'd use in T-SQL
 * the `offset` argument to `datepart` is in ticks-since-0001-01-01, so if you want to hard-code a minus-seven-hour offset, it looks like `-7h` and not `-7`

Would love your thoughts, and I'll update this discussion when we've posted a preview build (which will be in an hour or two, if I can get CI to agree with me :-)).

Cheers,
Nick

**Comment by nblumhardt** [comment]
2021.2.5483-pre is now on Downloads and Docker Hub, if you'd like to give it a try :-)
